### PR TITLE
add collective communication to generic vanguard

### PR DESCRIPTION
**Pull request by akva2**
@blattms this what you had in mind right.
@jcbowden this should get you going.

### CODE REVIEW

**Comment by blattms** [comment]
That looks about right. Using a unique-ptr seems a bit complicated. Can't we simply skip the pointer and use the communication object by value? The object should by quite slim and copying the MPI_Comm inside is not an issue.

**Comment by akva2** [comment]
Absolutely. I used unique_ptr to be consistent with other statics in vanguards, but I agree it give us absolutely nothing except complications here.

**Comment by jcbowden** [comment]
Hi,
Can akva2 get me a diff between release https://github.com/OPM/opm-simulators/tree/release/2021.04 and the changes in the akva2:add_collective_comm branch?
I'd like to try to test it and I am working on 2021.04 as master would not compile/run.
Thanks,
Josh

**Comment by jcbowden** [comment]
ok, hold that last request - I can find the changes in "Files changed" above!

**Comment by jcbowden** [comment]
Ok, the three files:
ebos/eclgenericvanguard.cc 
ebos/eclgenericvanguard.hh
opm/simulators/flow/Main.hpp

the first two do not exist in release opm-simulators/2021.04.

Where did they arrive from?

**Comment by atgeirr** [comment]
> Where did they arrive from?

There is a significant refactoring going on to reduce compilation times.

> master would not compile/run.

Let us rather solve this! You can post an issue, or write me an email.

**Comment by atgeirr** [comment]
jenkins build this please

**Comment by blattms** [comment]
As this is one commit @jcbowden can simply do:
```
git checkout release/2021.04
git checkout -b export-comm
git cherry-pick ed3a172a1e40f416bddae33f65d96a392277fbf9
```

**Comment by jcbowden** [comment]
Thanks Markus, I'll give that a try.

**Comment by blattms** [comment]
... and that will fail miserably.

Personally, I would advice against branching of the 2021.04 release as you will have to sort the issues twice...

**Comment by atgeirr** [comment]
Lots of errors, probably all alike:
```
terminate called after throwing an instance of 'Dune::ParallelError'
  what():  ParallelError [CollectiveCommunication:/usr/include/dune/common/parallel/mpicollectivecommunication.hh:153]: You must call MPIHelper::instance(argc,argv) in your main() function before using the MPI CollectiveCommunication!
```

**Comment by blattms** [comment]
MPI thinks that it was not initialized (`MPI_initialized` returns false)

**Comment by atgeirr** [comment]
I think we should get this in, so we can use it as a building block when refactoring the MPI_WORLD_COMM issues. But I believe it must be a unique_ptr not a value object, to defer construction. I think the reason for the error I saw before is that the value object is constructed at static init time, before MPI_Init() is called, using a unique_ptr should solve that.

**Comment by akva2** [comment]
rebased, using a unique_ptr again, tested "working".

**Comment by akva2** [comment]
jenkins build this please

**Comment by akva2** [comment]
jenkins build this please

**Comment by akva2** [comment]
jenkins build this please

File: ebos/eclgenericvanguard.hh [hunk 1]
```diff
@@ -260,6 +270,17 @@ public:
...
     { return parallelWells_; }
 
+    //! \brief Set global communication.
+    static void setCommunication(std::unique_ptr<CommunicationType> comm)
+    { comm_ = std::move(comm); }
+
+    //! \brief Obtain global communicator.
+    static CommunicationType& comm()
+    {
+        assert(comm);
```
Comment (atgeirr) [inline ebos/eclgenericvanguard.hh:280]: This assert looks like it asserts that the function exists.

**Comment by akva2** [comment]
jenkins build this please

**Comment by akva2** [comment]
jenkins build this please

**Comment by akva2** [comment]
jenkins build this please

**Comment by akva2** [comment]
jenkins build this please

**Comment by akva2** [comment]
okay, finally got all variants it seems. reasonably happy with this, the python simulators.hpp change is perhaps a little brittle but it is well defined so i don't think it's a biggie.

**Comment by atgeirr** [comment]
Thanks, this is good! Merging.

**Pull request merged**